### PR TITLE
[BEAM-5867] Operator translation based on operator name

### PR DESCRIPTION
**Pull request by mareksimunek**
Smarter why to choose operator translation based on operators name by introducing NameBaseTranslationProviderTest. Useful e.g. when we want Join operator which fits in memory translate with `BroadcastHashJoinTranslator`

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

File: sdks/java/extensions/euphoria/build.gradle [hunk 1]
```diff
@@ -34,9 +34,10 @@ dependencies {
     testCompile library.java.slf4j_api
     testCompile library.java.hamcrest_core
     testCompile library.java.hamcrest_library
-    testCompile library.java.mockito_core
+    testCompile library.java.powermock
```
Comment (dmvk) [inline sdks/java/extensions/euphoria/build.gradle]: can we stick with mockito?

**Comment by dmvk** [comment]
I think this heads the right direction. We need a convenient and flexible way for user to choose which translator should be used for an operator.

I'm thinking about having some kind of `composable translator provider`, that could compose of multiple translators (in first matching provider wins order.? It may be far more flexible, than setting a default translator in NamedTranslatorProvider (this implies, that there can be only one other provider alongside the named one).

@mareksimunek @VaclavPlajt any thoughts about this?

**Comment by mareksimunek** [comment]
@dmvk I think it should be solved in another ticket. You could also pass `composable translator provider` to NamedTranslatorProvider where the name matching will be first. Also you could later allow to not have set default translator and put this one in the composable one, which will comply with your idea.

**Comment by VaclavPlajt** [comment]
@dmvk I like the idea of chaining (anyhow) `TranslationProvider`s so one does not need to define translations of all the operators every time new `TranslationProvider` is implemented. Having composite provider is aligned with that. 

But it would require some more attention than originally anticipated in @mareksimunek scope of current work. So would also prefer to do it as different issue.

File: sdks/java/extensions/euphoria/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/translate/NameBasedTranslatorProvider.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+import org.apache.beam.sdk.extensions.euphoria.core.client.operator.base.Operator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Translation provider that selects name matching translation (operator name starts same as added
+ * {@link Builder#addShortNameTranslation(Class, OperatorTranslator, String)} translation). If no
+ * match, then its used {@link Builder#setDefaultTranslationProvider(TranslatorProvider)}.
+ */
+public class NameBasedTranslatorProvider implements TranslatorProvider {
```
[the remark this answers is not in the document]
Comment (mareksimunek) [reply sdks/java/extensions/euphoria/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/translate/NameBasedTranslatorProvider.java]: rephrased and class link little edited

File: website/src/documentation/sdks/euphoria.md [hunk 3]
```diff
@@ -343,7 +343,7 @@ Represents left join of two (left and right) datasets on given key producing sin
 // KV(3, "3+rat"), KV(0, "0+null"), KV(4, "4+duck"), KV(3, "3+cat"),
 // KV(3, "3+rat"), KV(1, "1+X")]
 ```
-Euphoria support performance optimization called 'BroadcastHashJoin' for the `LeftJoin`. User can indicate through previous operator's output hint `.output(SizeHint.FITS_IN_MEMORY)` that output `Dataset` of that operator fits in executors memory. And when the `Dataset` is used as right input, Euphoria will automatically translated `LeftJoin` as 'BroadcastHashJoin'. Broadcast join can be very efficient when joining between skewed datasets.
+Euphoria support performance optimization called 'BroadcastHashJoin' for the `LeftJoin`. User can indicate by operators name that right side of left join fits in executors memory (more [NameBasedTranslation](#namebasedtranslatorprovider)). Broadcast join can be very efficient when joining between skewed datasets.
```
[the remark this answers is not in the document]
Comment (mareksimunek) [reply website/src/documentation/sdks/euphoria.md]: edited

File: website/src/documentation/sdks/euphoria.md [hunk 4]
```diff
@@ -567,6 +567,19 @@ Dataset<SomeEventObject> timeStampedEvents =
 //Euphoria will now know event time for each event
 ```
 
+## TranslationProviders
```
[the remark this answers is not in the document]
Comment (mareksimunek) [reply website/src/documentation/sdks/euphoria.md]: included

**Comment by robertwb** [comment]
Looks like this needs a rebase.

**Comment by VaclavPlajt** [comment]
This is now obsolete by [already merged refactoring](https://github.com/apache/beam/pull/7019). @mareksimunek Can we close it?

**Pull request closed without merging**